File: implementations/fastapi/harness/rules/dockerfile_conventions.py
"""[21] dockerfile-conventions: Dockerfile multi-stage·HEALTHCHECK·.dockerignore (container.md)

Unlike the other rules, this one doesn't iterate over `py_files` — it reads
`<root>/Dockerfile` and `<root>/.dockerignore` directly as text. It checks container.md's
principles of "reduce the final image size via a multi-stage build" and "expose the
container's own health status via HEALTHCHECK," plus the `.dockerignore` convention that
keeps unnecessary files (`.git/`, `.venv/`, tests, etc.) out of the build context.

Checked items:
(a) 2 or more `FROM` lines — a multi-stage build
(b) A `HEALTHCHECK` instruction is present
(c) A `USER` instruction is present — runs as non-root
(d) `.dockerignore` exists in the same directory as the Dockerfile, and includes at least
    one minimal exclusion pattern (`.git`, a virtualenv directory, a test directory)
"""

from __future__ import annotations

import os
import re

from .common import RuleResult, failed, passed, skipped

FROM_RE = re.compile(r"^\s*FROM\s+\S+", re.MULTILINE)
HEALTHCHECK_RE = re.compile(r"^\s*HEALTHCHECK\b", re.MULTILINE)
USER_RE = re.compile(r"^\s*USER\s+\S+", re.MULTILINE)
REASONABLE_EXCLUDES = (".git", ".venv", "venv", "__pycache__", "tests", "test")


def _find_dockerfile(root: str) -> str | None:
    for candidate in ("Dockerfile", os.path.join("examples", "Dockerfile")):
        path = os.path.join(root, candidate)
        if os.path.isfile(path):
            return path
    return None


def check(root: str, py_files: list[str]) -> RuleResult:
    del py_files  # for signature uniformity — unused, since this rule reads Dockerfile/.dockerignore directly
    result = RuleResult("dockerfile-conventions")

    dockerfile_path = _find_dockerfile(root)
    if dockerfile_path is None:
        result.add(skipped("No Dockerfile"))
        return result

    dockerfile_dir = os.path.dirname(dockerfile_path)
    label = os.path.relpath(dockerfile_path, root)

    try:
        with open(dockerfile_path, encoding="utf-8") as fh:
            content = fh.read()
    except OSError as e:
        result.add(failed(label, f"Cannot read the Dockerfile: {e}"))
        return result

    from_count = len(FROM_RE.findall(content))
    if from_count >= 2:
        result.add(passed(f"{label} (a multi-stage build, {from_count} FROM lines)"))
    else:
        result.add(failed(label, f"Not a multi-stage build — {from_count} FROM line(s) (2+ required, container.md)"))

    if HEALTHCHECK_RE.search(content):
        result.add(passed(f"{label} (HEALTHCHECK present)"))
    else:
        result.add(failed(label, "No HEALTHCHECK instruction (container.md)"))

    if USER_RE.search(content):
        result.add(passed(f"{label} (a non-root USER is present)"))
    else:
        result.add(failed(label, "No USER instruction — the container runs as root (container.md)"))

    dockerignore_path = os.path.join(dockerfile_dir, ".dockerignore")
    dockerignore_label = os.path.relpath(dockerignore_path, root)
    if not os.path.isfile(dockerignore_path):
        result.add(failed(dockerignore_label, "No .dockerignore file"))
        return result

    try:
        with open(dockerignore_path, encoding="utf-8") as fh:
            ignore_content = fh.read()
    except OSError as e:
        result.add(failed(dockerignore_label, f"Cannot read .dockerignore: {e}"))
        return result

    if any(pattern in ignore_content for pattern in REASONABLE_EXCLUDES):
        result.add(passed(f"{dockerignore_label} (confirmed a reasonable exclusion pattern)"))
    else:
        result.add(
            failed(
                dockerignore_label,
                ".dockerignore has no reasonable exclusion pattern such as .git/venv/tests",
            )
        )

    return result

File: implementations/go/harness/dockerfile_conventions.go
package main

import (
	"os"
	"path/filepath"
	"regexp"
	"strconv"
	"strings"
)

var (
	fromLine        = regexp.MustCompile(`(?mi)^\s*FROM\s+\S+`)
	healthcheckLine = regexp.MustCompile(`(?mi)^\s*HEALTHCHECK\b`)
	userLine        = regexp.MustCompile(`(?mi)^\s*USER\s+\S+`)
)

// dockerignoreExcludes — the minimum set of excludes required by
// container.md/local-dev.md. The source repository (.git) and secrets
// (.env*) are the two things that must never end up in the image, so they
// serve as the minimum bar for "reasonable excludes".
var dockerignoreExcludes = []string{".git", ".env"}

// checkDockerfileConventions — [18] dockerfile-conventions: reads examples/Dockerfile
// and its sibling .dockerignore directly and checks them (container.md). Unlike
// the other rules, this one does not recursively walk the directory tree — it
// is a single-file check against exactly two files (<root>/Dockerfile,
// <root>/.dockerignore), not Go source.
//
// (a) multi-stage build — 2 or more FROM lines
// (b) a HEALTHCHECK instruction is present
// (c) .dockerignore exists and includes reasonable excludes like .git/.env*
func checkDockerfileConventions(root string) RuleResult {
	result := RuleResult{Section: "dockerfile-conventions"}
	dockerfilePath := filepath.Join(root, "Dockerfile")
	content, err := os.ReadFile(dockerfilePath)
	if os.IsNotExist(err) {
		result.Findings = append(result.Findings, skipFinding("Dockerfile not found"))
		return result
	}
	if err != nil {
		result.Findings = append(result.Findings, failFinding("Dockerfile", "read failed: "+err.Error()))
		return result
	}
	src := string(content)

	fromCount := len(fromLine.FindAllString(src, -1))
	if fromCount >= 2 {
		result.Findings = append(result.Findings, passFinding("Dockerfile (multi-stage, "+strconv.Itoa(fromCount)+" FROM lines)"))
	} else {
		result.Findings = append(result.Findings, failFinding("Dockerfile", "not a multi-stage build ("+strconv.Itoa(fromCount)+" FROM line(s)) — the build stage and the runtime stage must be separated (docs/architecture/container.md)"))
	}

	if healthcheckLine.MatchString(src) {
		result.Findings = append(result.Findings, passFinding("Dockerfile (HEALTHCHECK present)"))
	} else {
		result.Findings = append(result.Findings, failFinding("Dockerfile", "no HEALTHCHECK instruction (docs/architecture/container.md)"))
	}

	if userLine.MatchString(src) {
		result.Findings = append(result.Findings, passFinding("Dockerfile (non-root USER present)"))
	} else {
		result.Findings = append(result.Findings, failFinding("Dockerfile", "no USER instruction — the container runs as root (docs/architecture/container.md)"))
	}

	dockerignorePath := filepath.Join(root, ".dockerignore")
	diContent, diErr := os.ReadFile(dockerignorePath)
	switch {
	case os.IsNotExist(diErr):
		result.Findings = append(result.Findings, failFinding(".dockerignore", "no .dockerignore file — unnecessary files (.git, secrets, etc.) could end up in the build context (docs/architecture/container.md)"))
	case diErr != nil:
		result.Findings = append(result.Findings, failFinding(".dockerignore", "read failed: "+diErr.Error()))
	default:
		diSrc := string(diContent)
		var missing []string
		for _, exclude := range dockerignoreExcludes {
			if !strings.Contains(diSrc, exclude) {
				missing = append(missing, exclude)
			}
		}
		if len(missing) == 0 {
			result.Findings = append(result.Findings, passFinding(".dockerignore (.git/.env exclusion confirmed)"))
		} else {
			result.Findings = append(result.Findings, failFinding(".dockerignore", "missing reasonable excludes ("+strings.Join(missing, ", ")+")"))
		}
	}

	return result
}
